src/main/scala/com/fulcrumgenomics/bam/api/SamOrder.scala

```diff
@@ -0,0 +1,221 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2017 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.bam.api
+
+import htsjdk.samtools.SAMFileHeader.{GroupOrder, SortOrder}
+import htsjdk.samtools.util.Murmur3
+import htsjdk.samtools.{SAMFileHeader, SAMUtils}
+
+
+/** Trait for specifying BAM orderings. */
+sealed trait SamOrder extends Product {
+  type A <: Ordered[A]
+
+  /** Sets the appropriate fields in the SAM header for the order. */
+  def applyTo(header: SAMFileHeader): SAMFileHeader = {
+    header.setSortOrder(sortOrder)
+    header.setGroupOrder(groupOrder)
+    header.setAttribute("SS", subSort.map(ss => sortOrder.name() + ":" + ss).orNull)
+    header
+  }
+
+  /** Returns the sort order that should be set in the SAM header. */
+  def sortOrder : SortOrder
+
+  /** Returns the group order that should be set in the SAM header. None will cause GO to be unset. */
+  def groupOrder: GroupOrder
+
+  /** The subsort string to be placed *after* the sort order in the SS tag. */
+  def subSort: Option[String]
+
+  /** Function to generate the sort key for sorting records into this ordering. */
+  def sortkey: (SamRecord) => A
+}
+
+object SamOrder {
+  /** The set of all possible values. */
+  val values: Seq[SamOrder] = Seq(Coordinate, Queryname, Random, RandomQuery, TemplateCoordinate, Unsorted, Unknown)
```

Comment:
I recall some PR in dagr to get all the classes that extend a trait reflectively :P